src/EntityImporter.php

```diff
@@ -9,181 +9,238 @@
-					$badgeItems[] = $badge->getSerialization();
-				}
-			}
-		}
-
-		return $badgeItems;
-	}
-
-	private function getReferencedEntities( EntityDocument $entity ) {
-		$snaks = $entity->getStatements()->getAllSnaks();
-		$entities = array();
-
-		foreach( $snaks as $snak ) {
-			$entities[] = $snak->getPropertyId()->getSerialization();
-
-			if ( $snak instanceof PropertyValueSnak ) {
-				$value = $snak->getDataValue();
-
-				if ( $value instanceof EntityIdValue ) {
-					$entities[] = $value->getEntityId()->getSerialization();
-				}
-			}
-		}
-
-		return array_unique( $entities );
-	}
-
-	private function importBadgeItems( array $entities ) {
-		$badgeItems = $this->getBadgeItems( $entities );
-		$this->importEntities( $badgeItems, false );
-	}
+use Wikibase\Repo\WikibaseRepo;
+use DataValues\UnboundedQuantityValue;
+use Wikibase\DataModel\Entity\ItemId;
+
+class EntityImporter
+{
+
+    private $statementsImporter;
+
+    private $badgeItemUpdater;
+
+    private $apiEntityLookup;
+
+    private $entityStore;
+
+    private $entityMappingStore;
+
+    private $logger;
+
+    private $statementsCountLookup;
+
+    private $idParser;
+
+    private $importUser;
+
+    private $batchSize;
+
+    public function __construct(
+        StatementsImporter $statementsImporter,
+        BadgeItemUpdater $badgeItemUpdater,
+        ApiEntityLookup $apiEntityLookup,
+        EntityStore $entityStore,
+        ImportedEntityMappingStore $entityMappingStore,
+        StatementsCountLookup $statementsCountLookup,
+        LoggerInterface $logger
+    )
+    {
+        $this->statementsImporter = $statementsImporter;
+        $this->badgeItemUpdater = $badgeItemUpdater;
+        $this->apiEntityLookup = $apiEntityLookup;
+        $this->entityStore = $entityStore;
+        $this->entityMappingStore = $entityMappingStore;
+        $this->statementsCountLookup = $statementsCountLookup;
+        $this->logger = $logger;
+
+        $this->idParser = new BasicEntityIdParser();
+        $this->importUser = User::newFromId(0);
+        $this->batchSize = 10;
+    }
+
+    public function importEntities(array $ids, $importStatements = true)
+    {
+        $batches = array_chunk($ids, $this->batchSize);
+
+        $stashedEntities = array();
+
+        foreach ($batches as $batch) {
+            $entities = $this->apiEntityLookup->getEntities($batch);
+
+            if ($entities) {
+                $this->importBadgeItems($entities);
+            } else {
+                $this->logger->error('Failed to retrieve items for batch');
+            }
+
+            $stashedEntities = array_merge($stashedEntities, $this->importBatch($batch));
+        }
+
+        if ($importStatements === true) {
+            foreach ($stashedEntities as $entity) {
+                $referencedEntities = $this->getReferencedEntities($entity);
+                $this->importEntities($referencedEntities, false);
+
+                $entity_new = $entity;
```

Comment:
Is it really needed to clone the entity and do the changes? Why not changing them in the place? Sorry if it sounds stupid, I don't know the full context here.

Reply:
I'm not a good php programmer. I'm not sure if this is working because the coded iterates over the statements and they are changed. If you think this can work I will try ....

Reply:
hmm, good point. I think it can stay like this.